File: tunnel-server/src/main/java/com/tuuzed/tunnel/server/http/HttpServerChannelHandler.java
package com.tuuzed.tunnel.server.http;

import com.tuuzed.tunnel.common.interceptor.HttpRequestInterceptor;
import com.tuuzed.tunnel.common.logging.Logger;
import com.tuuzed.tunnel.common.logging.LoggerFactory;
import com.tuuzed.tunnel.common.proto.ProtoMessage;
import com.tuuzed.tunnel.common.proto.ProtoRequest;
import com.tuuzed.tunnel.common.util.HttpUtils;
import com.tuuzed.tunnel.server.internal.AttributeKeys;
import io.netty.buffer.ByteBuf;
import io.netty.buffer.Unpooled;
import io.netty.channel.Channel;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelHandlerContext;
import io.netty.channel.ChannelInboundHandlerAdapter;
import io.netty.handler.codec.http.HttpContent;
import io.netty.handler.codec.http.HttpRequest;
import io.netty.handler.codec.http.HttpResponse;
import org.jetbrains.annotations.NotNull;

import java.net.SocketAddress;

@SuppressWarnings("Duplicates")
public class HttpServerChannelHandler extends ChannelInboundHandlerAdapter {
    private static final Logger logger = LoggerFactory.getLogger(HttpServerChannelHandler.class);

    @NotNull
    private final HttpServer httpServer;
    @NotNull
    private final HttpRequestInterceptor httpRequestInterceptor;

    public HttpServerChannelHandler(
        @NotNull HttpServer httpServer,
        @NotNull HttpRequestInterceptor interceptor
    ) {
        this.httpServer = httpServer;
        this.httpRequestInterceptor = interceptor;
    }

    @Override
    public void channelActive(ChannelHandlerContext ctx) throws Exception {
        super.channelActive(ctx);
        logger.trace("channelActive: {}", ctx);
    }

    @Override
    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
        logger.trace("channelInactive: {}", ctx);
        final String vhost = ctx.channel().attr(AttributeKeys.VHOST).get();
        final Long sessionToken = ctx.channel().attr(AttributeKeys.SESSION_TOKEN).get();
        if (vhost != null && sessionToken != null) {
            final HttpTunnelDescriptor descriptor = httpServer.getDescriptorByVhost(vhost);
            if (descriptor != null) {
                final long tunnelToken = descriptor.tunnelSessions().tunnelToken();
                final Channel tunnelChannel = descriptor.tunnelSessions().tunnelChannel();
                tunnelChannel.writeAndFlush(
                    new ProtoMessage(
                        ProtoMessage.Type.REMOTE_DISCONNECT,
                        Unpooled.copyLong(tunnelToken, sessionToken).array(),
                        null
                    )
                );
            }
            ctx.channel().attr(AttributeKeys.VHOST).set(null);
            ctx.channel().attr(AttributeKeys.SESSION_TOKEN).set(null);
        }
        super.channelInactive(ctx);
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
    }

    @Override
    public void exceptionCaught(ChannelHandlerContext ctx, Throwable cause) throws Exception {
        logger.trace("exceptionCaught: {}", ctx, cause);
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
    }

    @Override
    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
        if (msg instanceof HttpRequest) {
            channelReadHttpRequest(ctx, (HttpRequest) msg);
        } else if (msg instanceof HttpContent) {
            channelReadHttpContent(ctx, (HttpContent) msg);
        }
    }

    /**
     * 处理读取到的HttpRequest类型的消息
     */
    private void channelReadHttpRequest(ChannelHandlerContext ctx, HttpRequest msg) throws Exception {
        final String vhost = HttpUtils.getVhost(msg);
        if (vhost == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
            return;
        }
        ctx.channel().attr(AttributeKeys.VHOST).set(vhost);
        final HttpTunnelDescriptor descriptor = httpServer.getDescriptorByVhost(vhost);
        if (descriptor == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
            ctx.channel().attr(AttributeKeys.PASS).set(false);
            return;
        }
        ctx.channel().attr(AttributeKeys.PASS).set(true);
        final ProtoRequest protoRequest = descriptor.tunnelSessions().protoRequest();
        final SocketAddress localAddress = ctx.channel().localAddress();
        final SocketAddress remoteAddress = ctx.channel().remoteAddress();

        final HttpResponse httpResponse = httpRequestInterceptor.handleHttpRequest(
            localAddress, remoteAddress,
            protoRequest, msg
        );
        if (httpResponse != null) {
            final ByteBuf byteBuf = HttpUtils.httpResponseToBytes(httpResponse);
            ctx.channel().writeAndFlush(byteBuf);
            return;
        }
        final long sessionToken = descriptor.tunnelSessions().putSessionChannel(ctx.channel());
        ctx.channel().attr(AttributeKeys.SESSION_TOKEN).set(sessionToken);
        final long tunnelToken = descriptor.tunnelSessions().tunnelToken();
        final byte[] requestBytes = HttpUtils.httpRequestToBytes(msg);
        descriptor.tunnelSessions().tunnelChannel().writeAndFlush(
            new ProtoMessage(ProtoMessage.Type.TRANSFER,
                Unpooled.copyLong(tunnelToken, sessionToken).array(),
                requestBytes
            )
        );
    }

    /**
     * 处理读取到的HttpContent类型的消息
     */
    private void channelReadHttpContent(ChannelHandlerContext ctx, HttpContent msg) throws Exception {
        final Boolean pass = ctx.channel().attr(AttributeKeys.PASS).get();
        if (!(pass != null && pass)) { // 如果没有放行
            return;
        }
        final String vhost = ctx.channel().attr(AttributeKeys.VHOST).get();
        final Long sessionToken = ctx.channel().attr(AttributeKeys.SESSION_TOKEN).get();
        if (vhost == null || sessionToken == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
            return;
        }

        final HttpTunnelDescriptor descriptor = httpServer.getDescriptorByVhost(vhost);
        if (descriptor == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE);
            return;
        }

        final long tunnelToken = descriptor.tunnelSessions().tunnelToken();
        final byte[] contentBytes = new byte[msg.content().readableBytes()];
        msg.content().readBytes(contentBytes);

        descriptor.tunnelSessions().tunnelChannel().writeAndFlush(
            new ProtoMessage(ProtoMessage.Type.TRANSFER,
                Unpooled.copyLong(tunnelToken, sessionToken).array(),
                contentBytes
            )
        );
    }

}

File: tunnel2-server/src/main/kotlin/tunnel2/server/http/HttpServerChannelHandler.kt
package tunnel2.server.http

import io.netty.buffer.Unpooled
import io.netty.channel.ChannelFutureListener
import io.netty.channel.ChannelHandlerContext
import io.netty.channel.ChannelInboundHandlerAdapter
import io.netty.handler.codec.http.HttpContent
import io.netty.handler.codec.http.HttpRequest
import tunnel2.common.logging.LoggerFactory
import tunnel2.common.proto.ProtoCw
import tunnel2.common.proto.ProtoMessage
import tunnel2.server.interceptor.HttpRequestInterceptor
import tunnel2.server.internal.*

class HttpServerChannelHandler(
    private val registry: HttpRegistry,
    private val interceptor: HttpRequestInterceptor
) : ChannelInboundHandlerAdapter() {

    companion object {
        private val logger = LoggerFactory.getLogger(HttpServerChannelHandler::class.java)
    }

    @Throws(Exception::class)
    override fun channelActive(ctx: ChannelHandlerContext) {
        super.channelActive(ctx)
        logger.trace("channelActive: {}", ctx)
    }

    @Throws(Exception::class)
    override fun channelInactive(ctx: ChannelHandlerContext) {
        logger.trace("channelInactive: {}", ctx)
        val host = ctx.channel().attr<String>(AK_HOST).get()
        val sessionId = ctx.channel().attr<Long>(AK_SESSION_ID).get()
        if (host != null && sessionId != null) {
            registry.getDescriptorByHost(host)?.also {
                it.sessionChannels.tunnelChannel.writeAndFlush(
                    ProtoMessage(
                        ProtoCw.REMOTE_DISCONNECT,
                        it.sessionChannels.tunnelId,
                        sessionId
                    )
                )
            }
            ctx.channel().attr<String>(AK_HOST).set(null)
            ctx.channel().attr<Long>(AK_SESSION_ID).set(null)
        }
        super.channelInactive(ctx)
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
    }

    @Throws(Exception::class)
    override fun exceptionCaught(ctx: ChannelHandlerContext, cause: Throwable) {
        logger.trace("exceptionCaught: {}", ctx, cause)
        ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
    }

    @Throws(Exception::class)
    override fun channelRead(ctx: ChannelHandlerContext, msg: Any) {
        if (msg is HttpRequest) {
            channelReadHttpRequest(ctx, msg)
        } else if (msg is HttpContent) {
            channelReadHttpContent(ctx, msg)
        }
    }

    /**处理读取到的HttpRequest类型的消息 */
    @Throws(Exception::class)
    private fun channelReadHttpRequest(ctx: ChannelHandlerContext, msg: HttpRequest) {
        val host = msg.host()
        if (host == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
            return
        }
        ctx.channel().attr<String>(AK_HOST).set(host)
        val descriptor = registry.getDescriptorByHost(host)
        if (descriptor == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
            ctx.channel().attr<Boolean>(AK_PASSED).set(false)
            return
        }
        ctx.channel().attr<Boolean>(AK_PASSED).set(true)
        val httpResponse = interceptor.handleHttpRequest(
            ctx.channel().localAddress(),
            ctx.channel().remoteAddress(),
            descriptor.sessionChannels.tunnelRequest,
            msg
        )
        if (httpResponse != null) {
            ctx.channel().writeAndFlush(httpResponse.toBytes())
            return
        }
        val sessionId = descriptor.sessionChannels.putSessionChannel(ctx.channel())
        ctx.channel().attr<Long>(AK_SESSION_ID).set(sessionId)
        val tunnelId = descriptor.sessionChannels.tunnelId
        val requestBytes = msg.toBytes()
        descriptor.sessionChannels.tunnelChannel.writeAndFlush(
            ProtoMessage(
                ProtoCw.TRANSFER,
                tunnelId,
                sessionId,
                requestBytes
            )
        )
    }

    /** 处理读取到的HttpContent类型的消息 */
    @Throws(Exception::class)
    private fun channelReadHttpContent(ctx: ChannelHandlerContext, msg: HttpContent) {
        val passed = ctx.channel().attr<Boolean>(AK_PASSED).get() ?: return
        if (!passed) return
        val host = ctx.channel().attr<String>(AK_HOST).get()
        val sessionId = ctx.channel().attr<Long>(AK_SESSION_ID).get()
        if (host == null || sessionId == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
            return
        }
        val descriptor = registry.getDescriptorByHost(host)
        if (descriptor == null) {
            ctx.channel().writeAndFlush(Unpooled.EMPTY_BUFFER).addListener(ChannelFutureListener.CLOSE)
            return
        }

        val tunnelId = descriptor.sessionChannels.tunnelId
        val contentBytes = ByteArray(msg.content().readableBytes())
        msg.content().readBytes(contentBytes)

        descriptor.sessionChannels.tunnelChannel.writeAndFlush(
            ProtoMessage(
                ProtoCw.TRANSFER,
                tunnelId,
                sessionId,
                contentBytes
            )
        )
    }

}